BrainPortal/app/controllers/data_providers_controller.rb

```diff
@@ -115,22 +115,44 @@ def new_personal #:nodoc:
     @provider         = UserkeyFlatDirSshDataProvider.new( :user_id   => current_user.id,
                                                            :group_id  => provider_group_id,
                                                            :online    => true,
-                                                           :read_only => false
+                                                           :read_only => false,
                                                          )
     @groups           = current_user.assignable_groups
   end
 
-  # create by normal user,  only UserkeyFlatDirSshDataProvider
+  # Create by normal user, only UserkeyFlatDirSshDataProvider
+  # S3DataProvider, S3FlatDataProvider, S3MultiLevelDataProvider
   def create_personal
     normal_params = params.require_as_params(:data_provider)
                           .permit(:name, :description, :group_id,
                                   :remote_user, :remote_host,
-                                  :remote_port, :remote_dir
+                                  :remote_port, :remote_dir,
+                                  :type,
+                                  :cloud_storage_client_identifier,
+                                  :cloud_storage_client_token,
+                                  :cloud_storage_client_bucket_name,
+                                  :cloud_storage_client_path_start,
+                                  :cloud_storage_endpoint,
+                                  :cloud_storage_region,
                                  )
+
+    authorized_type = [UserkeyFlatDirSshDataProvider, S3DataProvider, S3FlatDataProvider, S3MultiLevelDataProvider]
+    dp_type         = normal_params[:type] || "UserkeyFlatDirSshDataProvider"
+
+    # Check if the type is an authorized class
+    if ! authorized_type.include?(dp_type.constantize)
+      respond_to do |format|
+        format.html { render :action => :new_personal}
```

Comment:
If the type is wrong, you just redirect to the NEW page without any message. Add a message to :base please

Reply:
You still haven't added an active record error to :base